Add support for JSON Web Token (JWT) as Authorization Grant and update the token's `expires` on refresh

pbuyle opened the pull request:
Add support for JSON Web Token (JWT) as Authorization Grant as specified in https://tools.ietf.org/html/draft-ietf-oauth-jwt-bearer-12.

Test suite has been added, but only for node since I've not been able to get browser tests to run.

blakeembrey commented:
Looks awesome! I was refactoring the browser tests today to remove the duplicate tests, so I'll get in tomorrow. Any idea what was failing in the browser though?

pbuyle commented:
I have not looked into the issue. Running the browser test provudes the following outputs:
```
[15:19:04] Using gulpfile ~/src/js-client-oauth2/gulpfile.js
[15:19:04] Starting 'test:browser'...
INFO [karma]: Karma v0.12.31 server started at http://localhost:9876/
INFO [launcher]: Starting browser Chrome
INFO [launcher]: Starting browser Firefox
INFO [launcher]: Starting browser PhantomJS
ERROR [launcher]: Cannot start PhantomJS
	
ERROR [launcher]: Cannot start Firefox
	Can not find the binary firefox
	Please set env variable FIREFOX_BIN
ERROR [launcher]: Cannot start Chrome
	Can not find the binary google-chrome
	Please set env variable CHROME_BIN
INFO [launcher]: Trying to start PhantomJS again (1/2).
ERROR [launcher]: Cannot start PhantomJS
	
INFO [launcher]: Trying to start PhantomJS again (2/2).
ERROR [launcher]: Cannot start PhantomJS
	
ERROR [launcher]: PhantomJS failed 2 times (cannot start). Giving up.
undefined
[15:19:05] 'test:browser' errored after 474 ms
[15:19:05] Error: 1
  at formatError (/home/vagrant/src/js-client-oauth2/node_modules/gulp/bin/gulp.js:169:10)
  at Gulp.<anonymous> (/home/vagrant/src/js-client-oauth2/node_modules/gulp/bin/gulp.js:195:15)
  at Gulp.emit (events.js:129:20)
  at Gulp.Orchestrator._emitTaskDone (/home/vagrant/src/js-client-oauth2/node_modules/gulp/node_modules/orchestrator/index.js:264:8)
  at /home/vagrant/src/js-client-oauth2/node_modules/gulp/node_modules/orchestrator/index.js:275:23
  at finish (/home/vagrant/src/js-client-oauth2/node_modules/gulp/node_modules/orchestrator/lib/runTask.js:21:8)
  at cb (/home/vagrant/src/js-client-oauth2/node_modules/gulp/node_modules/orchestrator/lib/runTask.js:29:3)
  at removeAllListeners (/home/vagrant/src/js-client-oauth2/node_modules/karma/lib/server.js:220:7)
  at Server.<anonymous> (/home/vagrant/src/js-client-oauth2/node_modules/karma/lib/server.js:231:9)
  at Server.g (events.js:199:16)
  at Server.emit (events.js:129:20)
  at net.js:1389:10
  at process._tickDomainCallback (node.js:403:11)
```

blakeembrey commented:
Interesting. It should be ok since the APIs are consistent, the merged tests will fix that. Does Chrome/Firefox/PhantomJS exist?

pbuyle commented:
I've added another 'feature' to my branch. When using the `refresh` method of a `ClientOAuth2Token` instance, its `accessToken` and `refreshToken` properties were updated, but not its `expires` property. I added the update of this property.

blakeembrey commented on client-oauth2.js:
```diff
@@ -703,6 +711,59 @@
+
+    // Authentication of the client is optional, as described in Section 3.2.1 of OAuth 2.0 [RFC6749]
+    if (options.clientId) {
+      headers['Authorization'] = 'Basic ' + btoa(options.clientId + ':' + options.clientSecret || '');
```
What's the `||` meant to do? It'll never run that, maybe you were meant to wrap the client secret like `(options.clientSecret || '')`?

pbuyle replied:
Yes, I meant the wrap the client secret like `(options.clientSecret || '')`, because a client secret is not required for a public client. If I remember correctly, a empty password is acceptable in a HTTP basic authentication header. So this would allow public client ot use the library.

blakeembrey replied:
I'm pretty sure you're correct on that. Maybe we should fix all of these later so they won't ever say "undefined"? It's out of scope though, so this fix will be cool :+1:
